File: src/main/java/gregtechlite/gtlitecore/api/module/Module.java
package gregtechlite.gtlitecore.api.module;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface Module
{

    /**
     * Id of the module, must be unique within its container (containerId).
     */
    String moduleId();

    /**
     * Id of the container (ModuleContainer) to associate the module with.
     */
    String containerId();

    /**
     * Name of the module, used human-readable name.
     */
    String name();

    /**
     * A list of ModIds that the module depends on. If any mods specified
     * are not present, the module will not load.
     */
    String[] modDependencies() default {};

    /**
     * Whether the module is the core module for its container. Each container
     * must have exactly one core module which will be loaded before all other
     * modules in the container. Core modules should not have mod dependencies
     * because this module used to loaded mod structure from API to mod main.
     */
    boolean isCore() default false;

    /**
     * Author of the module.
     */
    String author() default "";

    /**
     * Version of the module.
     */
    String version() default "";

    /**
     * A description of the module in the module configuration file.
     */
    String descriptions() default "";

}

File: src/main/kotlin/gregtechlite/gtlitecore/api/module/Module.kt
package gregtechlite.gtlitecore.api.module

/**
 * Marks a class as a module, discovered and configured by [ModuleManager].
 *
 * @param moduleId        The id of the module, must be unique within its container (`containerId`).
 * @param containerId     The id of the container to associate the module with.
 * @param name            The human-readable name of the module, different with `moduleId` format.
 * @param modDependencies All mod id that the module depends on. If any mods specified are not present,
 *                        the module will not load.
 * @param isCore          Whether the module is core for its container. Each container must have exactly
 *                        one core module which will be loaded before all other modules in the container.
 *                        Core modules should not have mod dependencies because it used to loaded mod
 *                        structure from API to mod main.
 * @param author          The author of the module.
 * @param version         The version of the module.
 * @param descriptions    The description of the module in the module configuration file.
 */
@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
annotation class Module(val moduleId: String,
                        val containerId: String,
                        val name: String,
                        val modDependencies: Array<String> = [],
                        val isCore: Boolean = false,
                        val author: String = "",
                        val version: String = "",
                        val descriptions: String = "")
